File: packages/react-native-vision-camera-text-recognition/android/src/main/java/com/margelo/nitro/camera/textrecognition/RecognizedTextMapper.kt
package com.margelo.nitro.camera.textrecognition

import android.graphics.Point as AndroidPoint
import android.graphics.Rect as AndroidRect
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.Text.Element as MLTextElement
import com.google.mlkit.vision.text.Text.Line as MLTextLine
import com.google.mlkit.vision.text.Text.TextBlock as MLTextBlock

private fun AndroidRect?.toRect(): Rect {
  val box = this ?: return Rect(0.0, 0.0, 0.0, 0.0)
  return Rect(box.left.toDouble(), box.right.toDouble(), box.top.toDouble(), box.bottom.toDouble())
}

private fun Array<AndroidPoint>?.toPoints(): Array<Point> {
  val points = this ?: return emptyArray()
  return points
    .map { point -> Point(point.x.toDouble(), point.y.toDouble()) }
    .toTypedArray()
}

private fun String?.toLanguageCodes(): Array<String> {
  return if (isNullOrBlank()) emptyArray() else arrayOf(this)
}

private fun MLTextElement.toTextElement(): TextElement {
  return TextElement(
    text,
    boundingBox.toRect(),
    cornerPoints.toPoints(),
    recognizedLanguage.toLanguageCodes(),
  )
}

private fun MLTextLine.toTextLine(): TextLine {
  return TextLine(
    text,
    boundingBox.toRect(),
    cornerPoints.toPoints(),
    recognizedLanguage.toLanguageCodes(),
    elements.map { element -> element.toTextElement() }.toTypedArray(),
  )
}

private fun MLTextBlock.toTextBlock(): TextBlock {
  return TextBlock(
    text,
    boundingBox.toRect(),
    cornerPoints.toPoints(),
    recognizedLanguage.toLanguageCodes(),
    lines.map { line -> line.toTextLine() }.toTypedArray(),
  )
}

fun Text.toRecognizedText(): RecognizedText {
  return RecognizedText(
    text,
    textBlocks.map { block -> block.toTextBlock() }.toTypedArray(),
  )
}

File: packages/react-native-vision-camera-text-recognition/ios/RecognizedTextMapper.swift
//
//  RecognizedTextMapper.swift
//  VisionCameraTextRecognition
//

import CoreGraphics
import Foundation
import MLKitTextRecognition
import NitroModules

private typealias MLText = MLKitTextRecognition.Text
private typealias MLTextBlock = MLKitTextRecognition.TextBlock
private typealias MLTextLine = MLKitTextRecognition.TextLine
private typealias MLTextElement = MLKitTextRecognition.TextElement
private typealias NitroTextBlock = margelo.nitro.camera.textrecognition.TextBlock
private typealias NitroTextLine = margelo.nitro.camera.textrecognition.TextLine
private typealias NitroTextElement = margelo.nitro.camera.textrecognition.TextElement

private func toRect(_ frame: CGRect) -> Rect {
  return Rect(
    left: frame.origin.x,
    right: frame.origin.x + frame.size.width,
    top: frame.origin.y,
    bottom: frame.origin.y + frame.size.height)
}

private func toPoints(_ values: [NSValue]?) -> [Point] {
  guard let values else {
    return []
  }
  return values.map { value in
    let point = value.cgPointValue
    return Point(x: point.x, y: point.y)
  }
}

private func toLanguageCodes(_ languages: [TextRecognizedLanguage]) -> [String] {
  return languages.compactMap { language in
    guard let code = language.languageCode, !code.isEmpty else {
      return nil
    }
    return code
  }
}

extension MLTextElement {
  fileprivate func toNitroTextElement() -> NitroTextElement {
    return NitroTextElement(
      text: text,
      boundingBox: toRect(frame),
      cornerPoints: toPoints(cornerPoints),
      recognizedLanguages: toLanguageCodes(recognizedLanguages))
  }
}

extension MLTextLine {
  fileprivate func toNitroTextLine() -> NitroTextLine {
    return NitroTextLine(
      text: text,
      boundingBox: toRect(frame),
      cornerPoints: toPoints(cornerPoints),
      recognizedLanguages: toLanguageCodes(recognizedLanguages),
      elements: elements.map { $0.toNitroTextElement() })
  }
}

extension MLTextBlock {
  fileprivate func toNitroTextBlock() -> NitroTextBlock {
    return NitroTextBlock(
      text: text,
      boundingBox: toRect(frame),
      cornerPoints: toPoints(cornerPoints),
      recognizedLanguages: toLanguageCodes(recognizedLanguages),
      lines: lines.map { $0.toNitroTextLine() })
  }
}

extension MLText {
  func toRecognizedText() -> RecognizedText {
    return RecognizedText(
      text: text,
      blocks: blocks.map { $0.toNitroTextBlock() })
  }
}
